### Task 1. Create one pandas dataframe that combines all the data scraped from May 22, 2022 together. Drop rows with missing job titles and/or job descriptions. Use `spacy` to tokenize all the job titles included in the cleaned dataframe. For each job title, find all the nouns and all the adjectives in the title and get their lowercased lemmatized form. Use the reformatted nouns to construct a vocabulary set for this dataframe. How many unique nouns are there? Construct another vocabulary set using the reformatted adjectives. How many unique adjectives are there? What kind of different information do the nouns versus the adjectives reveal about the specific job? 

In [1]:
import os
import numpy as np
import pandas as pd
import spacy
import re
import glob
from tqdm import tqdm
nlp = spacy.load('en_core_web_sm')
os.chdir('/Users/[editted]/Documents/BC/SOCY7700/compSoc/homework 13/indeed_scraped_data')

In [5]:
May22= pd. DataFrame()
for file in glob.glob('*.csv'):
    read_file = pd.read_csv(file)
    May22= May22.append(read_file, ignore_index=True)
print(May22.columns)
cleaned_may22 = May22.dropna(subset=['lnks_job_title','lnks_job_description'])
cleaned_may22.to_csv('cleaned_may22.csv')
cleaned_may22 =pd.read_csv('cleaned_may22.csv')
print(cleaned_may22.shape)
job_df = cleaned_may22.dropna(subset=['lnks_job_title']).reset_index()

/var/folders/6j/g9631c1s65l0d3hl76jkmfc40000gn/T/ipykernel_16008/3560853125.py:4: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  May22= May22.append(read_file, ignore_index=True)
/var/folders/6j/g9631c1s65l0d3hl76jkmfc40000gn/T/ipykernel_16008/3560853125.py:4: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  May22= May22.append(read_file, ignore_index=True)
/var/folders/6j/g9631c1s65l0d3hl76jkmfc40000gn/T/ipykernel_16008/3560853125.py:4: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  May22= May22.append(read_file, ignore_index=True)
/var/folders/6j/g9631c1s65l0d3hl76jkmfc40000gn/T/ipykernel_16008/3560853125.py:4: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.

Index(['lnks_link', 'lnks_job_title', 'lnks_company', 'lnks_company_url',
       'lnks_company_location', 'lnks_job_description'],
      dtype='object')
(4803, 7)


In [6]:
all_nouns=[]
all_adjs=[]

for i in tqdm(range(len(job_df))):
    job_title = job_df.loc[i, 'lnks_job_title']
    for token in nlp(job_title):
        if token.pos_ == "NOUN":
            all_nouns.append(token.lemma_.lower())
        if token.pos_ == "ADJ":
            all_adjs.append(token.lemma_.lower())
print(all_nouns)
print(all_adjs)
print(len(all_nouns))
print(len(all_adjs))

100%|██████████████████████████████████████| 4803/4803 [00:31<00:00, 151.65it/s]

['contract', 'business', 'entry', 'business', 'purchasing', 'time', 'plant', 'manager', 'consultant', 'pilot', 'meat', 'lab', 'secretary', 'pharmacist', 'charge', 'meat', 'time', 'accountant', '|', 'living', 'fitter', 'lead', 'lab', 'counsel', 'planning', 'supervisor', 'safety', 'consultant', 'specialist', 'rate', 'training', 'shift', 'hvac', 'time', 'operation', 'technician', 'l2', 'collection', 'operation', 'aviation', 'support', 'technician', 'compliance', 'management', 'time', 'technician', 'day', 'hr', 'day', 'am', 'am', 'documentation', 'coordinator', 'mortgage', 'time', 'lead', 'consultant', 'pharmacist', 'hour', 'hour', 'pay', 'increase', 'blakeney', 'shift', 'manufacturing', 'trial', 'laboratory', 'class', 'supervisor', 'accounting', 'program', 'manager', 'surveillance', 'consultant', 'travel', 'nursing', 'event', 'maintenance', 'tax', 'supervisor', 'manager', 'surgery', 'pharmacist', 'product', 'security', 'consultant', 'robotic', 'operation', 'director', 'hr', 'time', 'time'

In [7]:
noun_with_freq = dict()

for i in tqdm(range(len(job_df))):
    job_title = job_df.loc[i, 'lnks_job_title']
    for token in nlp(job_title):
         if token.pos_ == "NOUN":
            noun_with_freq[token.lemma_.lower()] = noun_with_freq.get(token.lemma_.lower(), 0) + 1

noun_list = list(noun_with_freq.keys())
print('there are', len(noun_list), 'unique nouns among indeed job titles on May 22')

100%|██████████████████████████████████████| 4803/4803 [00:29<00:00, 162.05it/s]

there are 364 unique nouns among indeed job titles on May 22


In [8]:
adj_with_freq = dict()

for i in tqdm(range(len(job_df))):
    job_title = job_df.loc[i, 'lnks_job_title']
    for token in nlp(job_title):
         if token.pos_ == "ADJ":
            adj_with_freq[token.lemma_.lower()] = adj_with_freq.get(token.lemma_.lower(), 0) + 1

adj_list = list(adj_with_freq.keys())
print('there are', len(adj_list), 'unique adjectives among indeed job titles on May 22')

100%|██████████████████████████████████████| 4803/4803 [00:28<00:00, 167.97it/s]

there are 84 unique adjectives among indeed job titles on May 22


In [9]:
print(noun_list)
print(adj_list)

['contract', 'business', 'entry', 'purchasing', 'time', 'plant', 'manager', 'consultant', 'pilot', 'meat', 'lab', 'secretary', 'pharmacist', 'charge', 'accountant', '|', 'living', 'fitter', 'lead', 'counsel', 'planning', 'supervisor', 'safety', 'specialist', 'rate', 'training', 'shift', 'hvac', 'operation', 'technician', 'l2', 'collection', 'aviation', 'support', 'compliance', 'management', 'day', 'hr', 'am', 'documentation', 'coordinator', 'mortgage', 'hour', 'pay', 'increase', 'blakeney', 'manufacturing', 'trial', 'laboratory', 'class', 'accounting', 'program', 'surveillance', 'travel', 'nursing', 'event', 'maintenance', 'tax', 'surgery', 'product', 'security', 'robotic', 'director', 'tracking', 'food', 'safety/', 'inspector', 'ga', 'transportation', 'case', 'teacher', 'facility', 'production', 'quality', 'reporting', 'consumer', '-retail', 'service', 'office', 'benefit', 'rio', 'field', 'officer', 'information', 'afternoon', 'evening', 'marketing', 'housekeeper', 'oversight', 'engin

#### There are 364 unique nouns and 84 unique adjs. The number of nouns is much larger than the number of adjectives, which means there are many different job titles, but the requirements for these jobs are similar. For example, most jobs are full time or require licensed.

### Task 2. Choose the first job title in your dataframe as the primary string. Use one-hot encoding as the word embedding method and find jobs in your cleaned dataframe that have similar nouns in the title as your primary string.

In [10]:
vocab_with_freq = dict()

for i in tqdm(range(len(job_df))):
    job_title = job_df.loc[i, 'lnks_job_title']
    for token in nlp(job_title):
        if token.is_alpha:
            vocab_with_freq[token.lemma_.lower()] = vocab_with_freq.get(token.lemma_.lower(), 0) + 1

vocabulary_list = list(vocab_with_freq.keys())

100%|██████████████████████████████████████| 4803/4803 [00:28<00:00, 167.63it/s]


In [11]:
one_hot_encodings = []

for i in tqdm(range(len(job_df))):
    job_title = job_df.loc[i, 'lnks_job_title'] 
    token_indices = []

    for token in nlp(job_title):
        if token.is_alpha and token.lemma_.lower() in vocabulary_list:
            token_index_in_vocab = vocabulary_list.index(token.lemma_.lower())
            token_indices.append(token_index_in_vocab)
        
#         if token.is_alpha:
#             token_index_in_vocab = vocabulary_list.index(token.lemma_.lower())
#             token_indices.append(token_index_in_vocab)


    one_hot_encoding = np.zeros(len(vocabulary_list))
    for token_index in token_indices:
        one_hot_encoding[token_index] = 1

    one_hot_encodings.append(one_hot_encoding)

100%|██████████████████████████████████████| 4803/4803 [00:30<00:00, 159.96it/s]


In [12]:
np.array(one_hot_encodings).shape

(4803, 2047)

In [14]:
from scipy.spatial.distance import cosine
title_a = job_df.loc[0, 'lnks_job_title']
similarity_values = []
for i in tqdm(range(1, len(job_df))):
    similarity_value = 1 - cosine(one_hot_encodings[0], one_hot_encodings[i])
    similarity_values.append(similarity_value)

 32%|███████████▍                        | 1532/4802 [00:00<00:00, 15318.63it/s]/Users/apple/opt/anaconda3/lib/python3.9/site-packages/scipy/spatial/distance.py:699: RuntimeWarning: invalid value encountered in double_scalars
  dist = 1.0 - uv / np.sqrt(uu * vv)
100%|████████████████████████████████████| 4802/4802 [00:00<00:00, 15077.26it/s]


In [15]:
similar_df = pd.DataFrame(columns=['job_title', 'similarity_value_with_one_hot'])
similar_df['job_title'] = job_df.loc[1:, 'lnks_job_title']
similar_df['similarity_value_with_one_hot'] = similarity_values

print('the primary job title:', title_a)
similar_df.sort_values(by='similarity_value_with_one_hot', ascending=False)

the primary job title: Maintenance Controller (A&P) Technician


job_title  similarity_value_with_one_hot
2995                                     PCT1                            1.0
4471  Maintenance Controller (A&P) Technician                            1.0
1805  Maintenance Controller (A&P) Technician                            1.0
4786  Maintenance Controller (A&P) Technician                            1.0
2922  Maintenance Controller (A&P) Technician                            1.0
...                                       ...                            ...
1634                          Project Manager                            0.0
1633        CHBS-Mental Health Professional 2                            0.0
1632                   Eligibility Specialist                            0.0
1631                                      LPN                            0.0
4802                           Grants Manager                            0.0

[4802 rows x 2 columns]

### Task 3. Use spacy's word vector to do Task 2. Compare the results.

In [16]:
spacy_encoding = np.array([float(0) for i in range(96)])
token_count = 0

for i in tqdm(range(len(job_df))):
    job_title = job_df.loc[i, 'lnks_job_title']

    for token in nlp(job_title): 
        token_count += 1
        spacy_encoding += np.array(token.vector)
spacy_encoding /= token_count

spacy_encoding

100%|██████████████████████████████████████| 4803/4803 [00:27<00:00, 177.25it/s]


array([ 0.78727017,  0.09429036, -0.02919043, -0.30666952, -0.01297144,
        0.27987619, -0.66898868,  0.27417199,  0.15756936, -0.27501641,
        0.79000243, -0.29017478,  0.03133202, -0.15388452, -0.10432748,
       -0.21903943,  0.70736653,  0.36688579,  0.26753023,  0.0673385 ,
       -0.43341727, -0.59232369, -0.20853953,  0.17640638,  0.596861  ,
        0.1370358 , -0.36934622,  0.4998524 , -0.13412878,  0.71825226,
       -0.15560199,  0.16054755, -0.48123365, -0.37997171,  0.65542783,
       -0.06855987, -0.02637817,  0.29200202, -0.44460671, -0.06709029,
       -0.38061566,  0.11697892, -0.58253867, -0.7207017 ,  0.26194452,
       -0.994698  ,  0.3621341 ,  0.71141677,  0.22327731, -0.5851064 ,
        0.25521815, -0.03231519, -0.00201882, -1.15522795, -0.06106656,
        0.49936268,  0.70040712, -0.22699052, -0.58141122, -0.05059552,
       -0.18371167, -0.32731748,  0.53145735, -0.52224695,  0.60769446,
       -0.51141852,  0.07046536,  0.01111806,  0.27128419, -0.46

#### if we try to see the result directly, for example, we just want to get the similar words, then the method used in Q2 is much better and friendly for users. But the method used in Q3 is much better for researchers to do further research.